### DISCUSSION TITLE
Feature Request: Ability to run Cocotb tests without the simulator

### DISCUSSION THREAD

**Discussion by jshaker000** [cocotb/cocotb#2528] [Proposals]
I've been enjoying cocotb for a few months now. A feature that I think could be helpful would be the ability to perhaps have a python hook to run my tests without invoking the simulator at all.  
   
Would it be possible to use coroutines to have some framework to model HDL, or even just make a dummy DUT so I can check that I'm driving nets the way that I expect and I have no syntax issues and my test "generally works" without needing to elaborate and get a simulation license.  
   
I could imagine this feature essentially being two parts - having some function that could include a list of cocotb tests as a module or something and invoke them - annd then some way to make a 'dummy dut' object that a user could overload and pass to these tests.

Does that seem interesting or helpful - I'd be willing to expand on this idea if anyone finds this interesting. It is just a bit painful to find a python runtime error or an logical error in stimulus after spending a decent amount of time elaborating and loading the VPI and all ..

**Comment by jshaker000** [comment]
I guess part of the feasibility of this depends on how much a 'dut' object remembers or learns about itself or if it is essentially just a handle to talk to the simulator ..

It also depends on how the event model is handled in cocotb and if python 'knows' anything about the clock - Ie are these coroutines kind of blocking in python because they know the clock is going up and down or are they essentially looping through the simulator to stop and start.

I dont necessarily have a good implementation idea, but its been brewing in my head as something perhaps worth trying to support if its not too painful. I can imagine a lot of use of being able to run and analyze 'dry runs' of tests to check the python is right before running the dut - but I could also see as a user to be awkward to run against two versions. I'm sort of thinking out loud here.

**Comment by eric-wieser** [comment]
The danger here is that we end up reimplementing a simulator in python just to enable running without a simulator.

> without needing to elaborate and get a simulation license.

You could always use something like `iverilog` side-by-side with your "real" simulator, for testing cocotb without any interesting hardware.

> It also depends on how the event model is handled in cocotb and if python 'knows' anything about the clock - Ie are these coroutines kind of blocking in python because they know the clock is going up and down or are they essentially looping through the simulator to stop and start.

A coroutine that does `await RisingEdge(clock)` is waiting for the simulator to say "hey, a rising edge just happened on `clock`". Cocotb then finds and resumes the coroutine that was waiting on that. Once the coroutine pauses again, cocotb tells the simulator "ok I've handled that event, resume the simulation" by returning from the callback.

**Comment by jshaker000** [comment]
Right - definately implementing a sexond simulator in python is beyond the scope of cocotb and is definately not worth it.   
  
Perhaps the solution then is to in verilog simply stub out the guts of a module to see if the python is rougly right before elaborating the real thing. hmm

**Comment by ktbarrett** [comment]
Do you mind if I convert this into a discussion? Also, note that this is also discussed here https://github.com/cocotb/cocotb/pull/1332.

**Comment by ktbarrett** [comment]
> Would it be possible to use coroutines to have some framework to model HDL, or even just make a dummy DUT so I can check that I'm driving nets the way that I expect and I have no syntax issues and my test "generally works" without needing to elaborate and get a simulation license.

They call this mocking. I can imagine a use case where you just mock the interface when developing a driver, monitor, etc. without having a working HDL design to test against or loading a simulator. Since the dut object is just a nested object with basic functionality (get/set/iterate/index), this is something you can make yourself, and seeing how it is specific to a particular design/interface I don't see a way around it. Automating generation of these mock objects could be done, but I'm not sure if this needs to be in the main repo. Finally, the interface of the handles should be simplified to a point that it can actually be easily mocked.

The real problem for us starts with the scheduler and triggers. The triggers really should be refactored to dispatch to the dut objects for implementation. All simulation "bits" can and should be encapsulated in the handles (besides the concept of time/timesteps). The scheduler should already keep track of time step state to be able to check invalid operations (like `ReadWrite` after `ReadOnly`). Adding the concept of time on top of that shouldn't be difficult. For other scheduler bits (like `fork`), I think just refactoring the one cocotb has to be able to run outside of a simulator is feasible. @cmarqu has shown interest in this as well: the ability to use cocotb as a simulation library (in the classical meaning of "simulation" e.g. Simula-67 and various libraries in the same vain). 

> worth trying to support if its not too painful

All possible solutions will be very painful. That doesn't mean it isn't worth doing.

**Comment by ktbarrett** [comment]
Also, there is mypy for static checking which may prevent some simpler issues. I have not had any luck with it, but maybe you would. I think adding enough typing information to cocotb interfaces for a user to use mypy if they want is a good goal.

**Comment by jshaker000** [comment]
Making this a discussion is fine with me.

That explanation makes compete sense - users can write an arbitrary class that has the methods the dut has and most code would work (perhaps a cocotb geberated dut could have a method "real hardware" that against simulator returns true and otherwise returns false). I could also imagine cocotb have a 'discovery mode' where it makes a stub dut after running with the simulator onxe - but I thibk that thats secondary to the main issue and would be a utility. A user could easily write a mock model themselves.


But nonetheless there is the issue that scheduling is linked to the simulator so nothing would happen.

I think you outlined well why exactly Id want to use this

**Comment by ktbarrett** [comment]
There would have to be a runtime involved when mocking, but it could be made to be simple. The "test" would drive the mocked dut object which would cause edges to occur, the runtime would encode the concept of time, the dut would have `__rising_edge__` special methods that would be used for edges and would set the event that the trigger relied on, and so on...

It would be a simulator, but perhaps a simple one. The issue is that the cocotb codebase is in no shape to accept that currently. Partitioning of responsibility is very poor. The assumption of using an actual HDL simulator over the VPI/VHPI/FLI and *only* that is everywhere. It needs to kick up the level of abstraction a bit. 

Basically, redesign from the top down with this use case in mind. Design a well-defined interface. Fit the simulator interaction onto that. The current design is very hardcoded and "bottom-up", worrying more about implementation that interface.
